File: cwt-spring/src/main/java/com/cwtsite/cwt/domain/group/entity/Group.java
package com.cwtsite.cwt.domain.group.entity;

import com.cwtsite.cwt.entity.GroupStanding;
import com.cwtsite.cwt.domain.game.entity.Game;
import com.cwtsite.cwt.domain.group.entity.enumeration.GroupLabel;
import com.cwtsite.cwt.domain.tournament.entity.Tournament;

import javax.persistence.*;
import java.io.Serializable;
import java.util.List;
import java.util.Objects;

@Entity
@Table(name = "\"group\"")
@SequenceGenerator(name = "group_seq", sequenceName = "group_seq", initialValue = 121, allocationSize = 1)
public class Group implements Serializable {

    private static final long serialVersionUID = 1L;

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "group_seq")
    private Long id;

    @Enumerated(EnumType.STRING)
    @Column(name = "label")
    private GroupLabel label;

    @ManyToOne
    private Tournament tournament;

    @OneToMany(cascade = CascadeType.ALL, mappedBy = "group")
    private List<GroupStanding> standings;

    @OneToMany(cascade = CascadeType.ALL, mappedBy = "group")
    private List<Game> games;

    public Group() {
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public GroupLabel getLabel() {
        return label;
    }

    public void setLabel(GroupLabel label) {
        this.label = label;
    }

    public Group label(GroupLabel label) {
        this.label = label;
        return this;
    }

    public Tournament getTournament() {
        return tournament;
    }

    public void setTournament(Tournament tournament) {
        this.tournament = tournament;
    }

    public Group tournament(Tournament tournament) {
        this.tournament = tournament;
        return this;
    }

    public List<Game> getGames() {
        return games;
    }

    public void setGames(List<Game> games) {
        this.games = games;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Group group = (Group) o;
        if (group.id == null || id == null) {
            return false;
        }
        return Objects.equals(id, group.id);
    }

    public List<GroupStanding> getStandings() {
        return standings;
    }

    public void setStandings(List<GroupStanding> standings) {
        this.standings = standings;
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return "Group{" +
                "id=" + id +
                ", label='" + label + "'" +
                '}';
    }
}

File: cwt-spring/src/main/java/com/cwtsite/cwt/domain/group/entity/Group.kt
package com.cwtsite.cwt.domain.group.entity

import com.cwtsite.cwt.domain.group.entity.enumeration.GroupLabel
import com.cwtsite.cwt.domain.tournament.entity.Tournament
import com.cwtsite.cwt.entity.GroupStanding
import javax.persistence.Column
import javax.persistence.CascadeType
import javax.persistence.Entity
import javax.persistence.Enumerated
import javax.persistence.EnumType
import javax.persistence.GenerationType
import javax.persistence.GeneratedValue
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.SequenceGenerator
import javax.persistence.Table

@Entity
@Table(name = "\"group\"")
@SequenceGenerator(name = "group_seq", sequenceName = "group_id_seq", allocationSize = 1)
data class Group (

        @Id
        @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "group_seq")
        var id: Long? = null,

        @Enumerated(EnumType.STRING)
        @Column(name = "label")
        var label: GroupLabel? = null,

        @ManyToOne(optional = false)
        @JoinColumn(nullable = false)
        var tournament: Tournament? = null,

        @OneToMany(cascade = [CascadeType.ALL])
        @JoinColumn(name = "group_id")
        val standings: MutableList<GroupStanding> = mutableListOf()
) {

    override fun toString(): String = "Group{id=$id, labeel=$label}"

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Group

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id?.hashCode() ?: 0
    }
}
